Kotlin: Add option to add defaults to nullable fields

zmack opened the pull request:
---

### Description

This adds the option to have kotlin classes generated with constructors that set the default value of nullable types set to null.

This would end up generating classes to the tune of

```kotlin
data class Commit(
    override
    val abbreviatedOid: String,
    val additions: Int,
    val associatedPullRequests: PullRequestConnection? = null,
    val author: GitActor? = null,
...
```

rather than

```kotlin
data class Commit(
    override
    val abbreviatedOid: String,
    val additions: Int,
    val associatedPullRequests: PullRequestConnection?,
    val author: GitActor?,
...
```

This ends up being convenient in Kotlin as it makes using constructors more terse and to some degree obviates the need for generating builders.

Please let me know if I missed adding anything here.

Thank you.

---

Changes were made to:
- [ ] Codegen library - Java
- [x] Codegen library - Kotlin
- [ ] Codegen library - Scala
- [x] Maven plugin
- [x] Gradle plugin
- [ ] SBT plugin

jxnu-liguobin commented:
Welcome,   it looks good, Let's wait until the test passes. 👍🏻

zmack commented:
@jxnu-liguobin a bit confused as to why that test's failing, all of the constants seem to be in the right place, am I missing something ?

kobylynskyi commented on src/main/java/com/kobylynskyi/graphql/codegen/kotlin/KotlinGraphQLCodegen.java:
```diff
@@ -49,6 +49,9 @@ protected void initDefaultValues(MappingConfig mappingConfig) {
             mappingConfig.setGenerateModelOpenClasses(false);
         }
+        if (mappingConfig.isAutomaticallyDefaultNullable() == null) {
+            mappingConfig.setAutomaticallyDefaultNullable(true);
```
Here you define a default value of "true", while in `codegen-options.md` and in `MappingConfigConstants.java` you have false. This can confuse library users. Can you please clarify? Thanks

zmack replied:
Well spotted. Will update.

kobylynskyi commented:
@zmack please change the target branch to **develop** (according to [CONTRIBUTING.md#development-process](https://github.com/kobylynskyi/graphql-java-codegen/blob/master/CONTRIBUTING.md#development-process))
Thanks.

zmack commented on docs/codegen-options.md:
````diff
@@ -56,6 +56,7 @@ See [DirectiveAnnotationsMapping](#option-directiveannotationsmapping)* |
 | `responseProjectionMaxDepth`                          | Integer                                                               | 3                                                      | Sets max depth when use `all$()` which for facilitating the construction of projection automatically, the fields on all projections are provided when it be invoked. This is a global configuration, of course, you  ...
 | `generatedLanguage`                                   | Enum                                                                  | GeneratedLanguage.JAVA                                 | Choose which language you want to generate, Java,Scala,Kotlin were supported. Note that due to language features, there are slight differences in default values between languages.|
 | `generateModelOpenClasses`                            | Boolean                                                               | false                                                  | The class type of the generated model. If true, generate normal classes, else generate data classes. It only support in kotlin(```data class```) and scala(```case class```). Maybe we will consider to support Java ...
+| `automaticallyDefaultNullable`                        | Boolean                                                               | false                                                  | Adds a default null value to nullable arguments. Only supported in Kotlin.
````
Yup, good call, I'll rename it to that.

zmack commented on plugins/gradle/graphql-java-codegen-gradle-plugin/src/main/java/io/github/kobylynskyi/graphql/codegen/gradle/GraphQLCodegenGradleTask.java:
```diff
@@ -844,4 +846,8 @@ public Boolean isGenerateModelOpenClasses() {
+
+    public void setAutomaticallyDefaultNullable(Boolean automaticallyDefaultNullable) {
+        this.automaticallyDefaultNullable = automaticallyDefaultNullable;
+    }
```
Added

zmack commented:
> @zmack please change the target branch to **develop** (according to [CONTRIBUTING.md#development-process](https://github.com/kobylynskyi/graphql-java-codegen/blob/master/CONTRIBUTING.md#development-process))
> Thanks.

Oh welp, apologies, I totally missed that. Updated the target branch and rebased and that took care of all my testing woes 🥳 
Thank you.

kobylynskyi reviewed (approved):
@zmack looks good now!

kobylynskyi commented:
Nope, that's because Sonar run is not supported in PRs. You are good.

Djaler commented:
Sorry for necroposting, but do you intentionally not add default null [in parameterized input template](https://github.com/zmack/graphql-java-codegen/blob/9d8762de5a1760fcaa69002cd5eceb9d0b8c0c33/src/main/resources/templates/kotlin-lang/parametrized_input.ftl#L36) because of something, or this is a bug?
